src/native/libs/System.Native/pal_io.c

```diff
@@ -1325,7 +1325,7 @@ int32_t SystemNative_CopyFile(intptr_t sourceFd, intptr_t destinationFd, int64_t
     // Try copying data using a copy-on-write clone. This shares storage between the files.
     if (sourceLength != 0)
     {
-        while ((ret = ioctl(outFd, FICLONE, inFd)) < 0 && errno == EINTR);
+        while ((ret = ioctl(outFd, (int)FICLONE, inFd)) < 0 && errno == EINTR);
```

Comment:
@ayakael 
Why are you casting FICLONE to int? Second argument of ioctl function is unsigned long and not int. 
This caused ppc64le build error. Can you please revert this change?

Reply:
This was a proposal by @am11, as build failed without setting `Wno-sign-conversion`. See: https://github.com/dotnet/runtime/pull/76500#discussion_r1010819744. 

Here is the exact error:
```
/var/build/dotnet7/testing/dotnet7-stage0/src/dotnet-e6dd91c290b808f971a1ac69c2fb29395bbf1051/src/runtime/src/native/libs/System.Native/pal_io.c(1270,36): error GBBD8A67C: implicit conversion changes signedness: 'unsigned long' to 'int' [-Wsign-conversion] [/var/build/dotnet7/testing/dotnet7-stage0/src/dotnet-e6dd91c290b808f971a1ac69c2fb29395bbf1051/src/runtime/src/native/libs/build-native.proj]
/var/build/dotnet7/testing/dotnet7-stage0/src/dotnet-e6dd91c290b808f971a1ac69c2fb29395bbf1051/src/runtime/src/native/libs/System.Native/pal_io.c(1270,36): error GBBD8A67C: implicit conversion changes signedness: 'unsigned long' to 'int' [-Wsign-conversion] [/var/build/dotnet7/testing/dotnet7-stage0/src/dotnet-e6dd91c290b808f971a1ac69c2fb29395bbf1051/src/runtime/src/native/libs/build-native.proj]
```

Reply:
If you change last argument to `unsigned long` here: https://github.com/dotnet/runtime/blob/cbc8695ae0c8c2c2d1ac1fc4546d81e0967ef716/src/native/libs/System.Native/pal_io.c#L76 without this `(int)FICLONE` cast, does it fix both `linux-ppc64le` and `linux-musl-ppc64le` builds?

Reply:
That would be incorrect - the `_IOW` macro uses `sizeof` its last argument as a component of the resulting value, so changing that size from 4 to 8 will result in an incorrect value of `FICLONE`.

This issue seems to be caused by some difference in the definition of `ioctl` with musl.  In glibc, we have this:
```
extern int ioctl (int __fd, unsigned long int __request, ...);
```
Is this declared differently in musl?

Reply:
It is a ppc64le specific issue, where musl and glibc have different definitions. Rest of the platform matrix (including linux-musl-x64/arm64/arm etc..) is fine either way.

> That would be incorrect - the _IOW macro uses sizeof

I see, so the current definition is also incorrect; should be `size_t` instead of `int`, right?

Reply:
> It is a ppc64le specific issue, where musl and glibc have different definitions. Rest of the platform matrix (including linux-musl-x64/arm64/arm etc..) is fine either way.

The only difference between ppc64le and other platforms is the numerical value of `FICLONE` here.  On ppc64le, this value is `0x80049409` while on other platforms it is `0x40049409` due to a different definition of the `_IOW` macro.  This means on other platforms the value is the same interpreted as `int` and `unsigned long`, so there's no compiler warning.  On ppc64le, the value is negative interpreted as `int` and positive interpreted as `unsigned long`, that's why we get the warning. The type mismatch itself is currently present on all platforms.

> > That would be incorrect - the _IOW macro uses sizeof
> 
> I see, so the current definition is also incorrect; should be `size_t` instead of `int`, right?

No.  We must use `sizeof (int)`, i.e. 4, here; this flows into the construction of that value above.

Reply:
> This issue seems to be caused by some difference in the definition of `ioctl` with musl. In glibc, we have this:
> 
> ```
> extern int ioctl (int __fd, unsigned long int __request, ...);
> ```
> 
> Is this declared differently in musl?

Indeed this seems to be problem.  In the musl headers I see instead:
```
int ioctl (int, int, ...);
```
I guess as long as the two headers differ, there's probably no way to fix the source to work with both libcs, except for using some #ifdef ...